domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt

```diff
@@ -0,0 +1,105 @@
+
+/**
+ * Private mutable class that stores a grade as a fraction.
+ */
+private class MutableFractionGrade(var numerator: Int, var denominator: Int)
+
+/**
+ * Private class that computes the state of the user's performance at the end of a practice session.
+ * This class can exist across multiple training session instances, but calling code is responsible
+ * for ensuring it is properly reset.
+ */
+@Singleton
+internal
+class QuestionAssessmentCalculation @Inject constructor(
+  @ViewHintScorePenalty private val viewHintPenalty: Int,
+  @WrongAnswerScorePenalty private val wrongAnswerPenalty: Int,
+  @MaxScorePerQuestion private val maxScorePerQuestion: Int
+) {
+  private lateinit var skillIdList: List<String>
+  private lateinit var questionSessionMetrics: List<QuestionSessionMetrics>
+  private lateinit var totalScore: MutableFractionGrade
+  private lateinit var scorePerSkillMapping: MutableMap<String, MutableFractionGrade>
+
+  /** Initialize member fields. */
+  internal fun initialize(
+    skillIdList: List<String>,
+    questionSessionMetrics: List<QuestionSessionMetrics>
+  ) {
+    this.skillIdList = skillIdList
+    this.questionSessionMetrics = questionSessionMetrics
+    this.totalScore =
+      MutableFractionGrade(0, 0)
+    createScorePerSkillMapping()
+  }
+
+  /** Initialize the scorePerSkillMapping member field. */
+  private fun createScorePerSkillMapping() {
+    scorePerSkillMapping = mutableMapOf()
+    for (skillId in skillIdList) {
+      scorePerSkillMapping[skillId] = MutableFractionGrade(0, 0)
+    }
+  }
+
+  /** Calculate the user's overall score and score per skill for this practice session. */
+  private fun calculateScores() {
+    for (questionMetric in questionSessionMetrics) {
+      val totalHintsPenalty =
+        questionMetric.numberOfHintsUsed * viewHintPenalty
+      val totalWrongAnswerPenalty =
+        (questionMetric.numberOfAnswersSubmitted - 1) * wrongAnswerPenalty
+      var questionScore = 0
+      if (!questionMetric.didViewSolution) {
+        questionScore =
+          maxOf(
+            maxScorePerQuestion - totalHintsPenalty - totalWrongAnswerPenalty,
+            questionScore
+          )
+      }
+      this.totalScore.numerator += questionScore
+      this.totalScore.denominator += maxScorePerQuestion
+      for (linkedSkillId in questionMetric.question.linkedSkillIdsList) {
+        if (!scorePerSkillMapping.containsKey(linkedSkillId)) continue
+        scorePerSkillMapping[linkedSkillId]!!.numerator += questionScore
+        scorePerSkillMapping[linkedSkillId]!!.denominator += maxScorePerQuestion
+      }
+    }
+  }
+
+  /** Compute the overall score as well as the score and mastery per skill. */
+  internal fun computeAll(): UserAssessmentPerformance {
+    calculateScores()
+    // TODO(#3067): Create mastery calculations method
+
+    // Set up the return values
+    var finalScore = FractionGrade.newBuilder().apply {
+      numerator = totalScore.numerator.toDouble() / 10
```

Comment:
Why divide by 10? Is it because we try to have 10 questions in a practice session? Note that this might not always be the case (e.g. we might change that requirement, make it dynamic in the future, or have fewer than 10 due to insufficient questions to select from when generating the practice session--the last one is a potential issue that can happen today).

Reply:
I think my comment may have gotten lost, but I believe Chantel wants to display the score for each skill as a fraction, and Oppia web is calculating scores as fractions represented by real value numerator/denominators. I took the approach we had previously discussed and we're now using integers to represent the score numerators/denominators and penalties internally to avoid floating point error - I did this by multiplying the total score and penalties by a factor of 10. To stay consistent with Oppia web if Chantel chooses to display the numerator/denominator of each score, I'm dividing everything by a factor of 10 before it is returned.